src/adapter/src/rbac.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use crate::catalog::Catalog;
+use crate::command::Command;
+use crate::session::Session;
+use crate::AdapterError;
+use mz_sql::catalog::{CatalogRole, SessionCatalog};
+use mz_sql::plan::Plan;
+use std::error::Error;
+use std::fmt;
+use std::fmt::Formatter;
+
+/// Attributes that allow a role to execute certain plans.
+///
+/// Note: This is a subset of all role attributes used for privilege checks.
+#[derive(Debug)]
+pub enum Attribute {
+    CreateRole,
+    CreateDB,
+    CreateCluster,
+}
+
+impl Attribute {
+    /// Checks if a role has the privilege granted by the attribute.
+    fn check_role(&self, role: &(impl CatalogRole + ?Sized)) -> bool {
+        match self {
+            Attribute::CreateRole => role.create_role(),
+            Attribute::CreateDB => role.create_db(),
+            Attribute::CreateCluster => role.create_cluster(),
+        }
+    }
+}
+
+impl ToString for Attribute {
```

Comment:
I never know when to use one of the other, so `ToString` was a bit random. I've switched this to `Display`

Reply:
FWIW the right thing to do is implement `Display`, because `ToString` is [implemented for all types T that impl Display](https://doc.rust-lang.org/std/string/trait.ToString.html#impl-ToString-for-T)